### ISSUE TITLE
Removing duplicates while transforming.

### ISSUE THREAD

**Issue by pratikpparikh** [schibsted/jslt#16]
@larsga I have another unique problem in dealing with nested structures and need your opinion. I have duplicate reference node, I need to figure out how to private the duplicate from being created. I started building a global array and thinking that I need to check that array in flattening to remove duplicates. If you can give it a look and let me know what you think the best option here is then it will be great. Below is current input, expected output, and the current template.

Input

> 
> [
> {
> "chapters": [
> {
> "name": "a-f"
> },
> {
> "name": "b"
> }
> ]
> },
> {
> "chapters": [
> {
> "name": "a-f"
> },
> {
> "name": "e"
> }
> ]
> },
> {
> "chapters": [
> {
> "name": "l-p"
> },
> {
> "name": "m"
> }
> ]
> },
> {
> "chapters": [
> {
> "name": "v-z"
> },
> {
> "name": "v"
> }
> ]
> },
> {
> "chapters": [
> {
> "name": "v-z"
> },
> {
> "name": "z"
> }
> ]
> }
> ]
> 

Expected Output

> [
> {
> "name": "a-f"
> },
> {
> "name": "b",
> "parents": [
> {
> "name": "a-f"
> }
> ]
> },
> {
> "name": "e",
> "parents": [
> {
> "name": "a-f"
> }
> ]
> },
> {
> "name": "l-p"
> },
> {
> "name": "m",
> "parents": [
> {
> "name": "l-p"
> }
> ]
> },
> {
> "name": "v-z"
> },
> {
> "name": "v",
> "parents": [
> {
> "name": "v-z"
> }
> ]
> },
> {
> "name": "z",
> "parents": [
> {
> "name": "v-z"
> }
> ]
> }
> ]
> 

Current Template

> def flatten(arrayofarrays)
> if ($arrayofarrays)
> $arrayofarrays[0] + flatten($arrayofarrays[1 : ])
> else
> []
> 
> let globalTopNodeCache = [for (.) if(.chapters and size(.chapters) > 0) .chapters[0].name]
> 
> flatten([for (.)
> let chapterZero = if(.chapters and size(.chapters) > 0) .chapters[0] else ""
> [{"name" : $chapterZero .name}] + [for (.chapters[1 : ]) {"parents":[{"name":$chapterZero .name}],"name" : .name}]
> ])

**Comment by larsga** [comment]
This gets pretty close:

```
def unique(values)
  let obj = {for ($values) . : .}
  [for ($obj) .value]

let chapters = [for (.) .chapters]
let parents = [for (unique([for ($chapters) .[0].name])) {"name" : . }]
let children = [for ($chapters) {
  "name" : .[1].name,
  "parents" : {"name" : .[0].name}
}]

$parents + $children
```

The only thing is that here the order is different, because the parents come first, instead of being interleaved with the children.

The interleaving can be done by nesting and then using `flatten`, but it requires supporting `if` in `for`loops. Which is the same thing we needed in the previous issue.

```
def flatten(arrayofarrays)
  if ($arrayofarrays)
    $arrayofarrays[0] + flatten($arrayofarrays[1 : ])
  else
    []

def unique(values)
  let obj = {for ($values) . : .}
  [for ($obj) .value]

let input = (.)
let chapters = unique([for (.) .chapters[0].name])

flatten([for ($chapters) 
    let name = (.)
   
    [
    {"name" : .},
    [for ($input) if (.chapters[0].name == $name) {
      "name" : .chapters[1].name,
      "parents" : {"name" : .chapters[0].name}
    }]
  ]
])
```

**Comment by pratikpparikh** [comment]
I think for now the current version should work for me.  But have the support to do if support for interleaving is important.  Can you explain to me how the unique is being identified?

def unique(values)
  let obj = {for ($values) . : .} // is this creating set?
  [for ($obj) .value]

**Comment by larsga** [comment]
Yes, we create a JSON object on the second line where the values are used as the key. This causes duplicate keys to overwrite each other, so that when we extract them in the third line we get only the unique values.

I've thought a bit more about this, and probably `group by` is the best long-term solution. That would work as follows: you can add `group by <expr>` to a for loop. This will cause your for loop to loop over the following structure:

```
[
  {"key" : key1,
   "values" : obj1, obj2, obj3, ... },

  {"key" : key2,
   "values" : obj4, obj5, obj6, ... },

  ...
]
```

In other words, you'll loop over an array of objects where each object has the group by-value (key) and the set of values from the original `for (<expr>)` that produced that exact key.

With that the implementation becomes:

```
flatten([for (.)
  let key = .key
  [
    {"name" : $key},
    [for (.values) {
      "name" : .[1].name,
      "parents" : {"name" : $key}
    }]
  ]
group by .chapters[0].name])
```

**Comment by larsga** [comment]
Stage 0: that's what the `group by` does. Simply including the `group by` in the for expression makes you loop over that structure. That's basically how `group by` works. Maybe the [idea writeup](https://github.com/schibsted/jslt/blob/master/ideas/group-by.md) will help. 

But the values are not as you show them. Instead they are

```
{"key" : "a-f",
  "values" : [

{
"chapters": [
{
"name": "a-f"
},
{
"name": "b"
}
]
},
{
"chapters": [
{
"name": "a-f"
},
{
"name": "e"
}

]}
```

`group by` doesn't know which part of the object you want, so it keeps everything.

Stage 1 is correct, and I think it's clear how it happens.

**Comment by pratikpparikh** [comment]
@larsgaI am on the same page now.  How hard is it to implement group by and if support for for-loop?

**Comment by larsga** [comment]
If support for the for loop is implemented now. So you have a solution now.

Group by is easy to implement, but I want to research alternatives a little more before implementing it.

**Comment by larsga** [comment]
Some notes [here](https://github.com/schibsted/jslt/blob/master/ideas/group-by.md).

**Comment by pratikpparikh** [comment]
@larsga yes, thank you, I have been tracking the changes.  Have not fully consumed group by and order by thought process but will provide feedback as soon as I do.

**Comment by pratikpparikh** [comment]
@larsga very well documented sir.  Thank you.

1. I like xQuery, very easy on the eyes.
2. I love comprehensions, but users now have to learn the ordering which can prevent correct utilization.
3. macro seems to be a good compromise. So, I think my vote is to that at this point.

If i understanding everything correctly, I also think macro will provide us with more flexibility of implementation as you will only be introducing built-in function rather than language.  Please let me know what you think?

**Comment by pratikpparikh** [comment]
@larsga if you get time can you look at group-by, I tried to give it a shot by creating a macro but could not get it to work as intended.  I was trying to solve group by multiple items.

**Comment by zsyphon** [comment]
@larsga @pratikpparikh  HI guys I've run into a similar group by issue as well. Have you been able to make any progress on this? Thank you for any help or updates you can provide!

**Comment by larsga** [comment]
I've [investigated this a bit](https://github.com/schibsted/jslt/blob/master/ideas/group-by.md) and it turns out it's pretty complicated to do it the way that I've been planning (in "for").

Adding a simple `group-by` built-in macro would be easier, but I'm not sure that's the route I want to take. So I'm still undecided on this.

How urgent is this issue for you two?

**Comment by zsyphon** [comment]
@larsga Thank you for looking into the issue! It's pretty urgent. I could work with a macro. Thank you for your work again. It's really appreciated!!

**Comment by pratikpparikh** [comment]
@larsga I have a version of the macro, it is not what I would ideally like but I was waiting for you to be back in your working schedule.  So, I can discuss this with you by means of PR and maybe work together to get the macro to an ideal place if you want I can send the PR to you?

**Comment by larsga** [comment]
I'm not convinced a macro is the way to go, and ideally I would like more time to experiment and think before adding this feature. But I guess we also need something that works for people now.

My proposal is to add an experimental feature now. Something that's not in the documentation, and which is clearly and obviously experimental, but still workable and has good performance.

What do you guys think of the following:

```
import "http://jslt.schibsted.com/2018/experimental" as exp

exp:group-by( <get values to group>, <expression to produce key> )
```

So we introduce an experimental module that contains things which may be added to the language in the future, or which may be superseded by other functionality later. Either way we have something that works but which is clearly not officially in the language.

Thoughts?

**Comment by pratikpparikh** [comment]
@larsga @zsyphon I agree with the approach.

**Comment by zsyphon** [comment]
@larsga @pratikpparikh I like this too. Thank you again!

**Comment by pratikpparikh** [comment]
@larsga what are the next steps?

**Comment by pratikpparikh** [comment]
@larsga no need to be sorry, by contributing to this project and help others you are doing a great service to the community.  I hope you have a successful launch and all turns well at the home front.  Please let me know if I can help by clarifying, implementing or testing something for your changes.  any help I can provide please don't hesitate.

**Comment by pratikpparikh** [comment]
@larsga  how you are recovering well and might have the time to look at this?

**Comment by larsga** [comment]
I'll need to do this during the weekend. I have most of the design in my head, but haven't had time to type it in yet.

**Comment by pratikpparikh** [comment]
@larsga thank you please let me know if I can help in any way.

**Comment by larsga** [comment]
Another use case came up today. The user wants to turn:

```{
"elements": [{"e_1": {"exposed": "v_1",
                    "placement": "v_2",
                    "widget": "v_3"}
             },
            {"e_2": {"exposed": "v_3",
                    "placement": "v_4",
                    "widget": "v_5"}
            },
            {"e_3": {"exposed": "v_5",
                    "placement": "v_6",
                    "widget": "v_7"} 
            }]         
}```

into:

```
{
    "exposed": {v_1,v_3,v_5, ...},
    "placement": {v_2,v_4,v_6,...},
    "widget": {v_3,v_5,v_7,...}
}
```

`group_by` is necessary to solve this one, too.

Correct input to `group_by` can be produced with:

```
flatten([for (.elements)
  [for (.) .value]
])
```

**Comment by zsyphon** [comment]
Thank you for the use case @larsga !

**Comment by larsga** [comment]
I've now added the experimental module with the `group-by` macro. The last use case can now be solved with the following:

```
import \"http://jslt.schibsted.com/2018/experimental\" as exp  
exp:group-by(flatten(
  [for (.elements) 
    [for (.) array(.value)] 
  ]),
  .key, .value)
```

The input to `group-by` will be:

```
[ {
  "key" : "exposed",
  "value" : "v_1"
}, {
  "key" : "placement",
  "value" : "v_2"
}, 
 ...
```

The macro will then group by the `.key` and for each of them collect `.value`.

**Comment by larsga** [comment]
@pratikpparikh Please try out the code that's in master now, and if it does what you need we can release 0.1.4.

**Comment by larsga** [comment]
I'm wondering whether perhaps the implementation in the repo now is not ideal. The keys have to be strings now, which may not work for all users. If instead of producing:

```
{
  "key1" : [v1, v2, ...],
  "key2" : [v3, v4, ...],
  ...
}
```

the code should perhaps produce:

```
[
  {"key" : "key1", "values" : [v1, v2, ...]},
  {"key" : "key2", "values" : [v3, v4, ...]},
  ...
]
```

Thoughts?

**Comment by pratikpparikh** [comment]
@larsgaI am going to try this out today and report back.  Additionally, the current limit is that we can only do the group by with one key.  But i think that is okay for now.

**Comment by larsga** [comment]
What do you mean "we can only do the group by with one key"?

**Comment by pratikpparikh** [comment]
@larsga i mean we can only group by one key.  We can't send a set of keys to aggregate by. for i.e.

import \"http://jslt.schibsted.com/2018/experimental\" as exp  
exp:group-by(flatten(
  [for (.elements) 
    [for (.) array(.value)] 
  ]),
  [.key,.key1], .value)

So, in this case, it should group by .key first followed by .key1.

**Comment by pratikpparikh** [comment]
@larsga arrays of objects are for sure preferred.

**Comment by larsga** [comment]
I changed it now. So the output is now on the form:

```
[{"key" : "key1", "values" : [v1,v4...]},
 {"key" : "key12, "values" : [v3,v4...]},
 ...]
```

Please test and check if you think it works well now. If you do we can release 0.1.4.

**Comment by pratikpparikh** [comment]
@larsga i will test this out.

**Comment by larsga** [comment]
Thanks!

**Comment by joaopastor** [comment]
hey there, I've been testing the group-by function, it seems to work ok.

Input:
```json
[
  {"key":643,"value":329511},
  {"key":643,"value":331997},
  {"key":643,"value":334182},
  {"key":643,"value":335042},
  {"key":643,"value":336741},
  {"key":643,"value":340372},
  {"key":642,"value":334029},
  {"key":642,"value":335042},
  {"key":642,"value":335156},
  {"key":642,"value":350178},
  {"key":642,"value":428917},
  {"key":642,"value":501998},
  {"key":642,"value":569184},
  {"key":642,"value":1040637},
  {"key":642,"value":1325719},
  {"key":642,"value":1358926},
  {"key":642,"value":1470467},
  {"key":642,"value":1470468}
]
```

Transformation:
```
import "http://jslt.schibsted.com/2018/experimental" as exp
exp:group-by(., .key, .value)
```

Output:
```json
[
  {
    "key":642,
    "values":[
      334029,
      335042,
      335156,
      350178,
      428917,
      501998,
      569184,
      1040637,
      1325719,
      1358926,
      1470467,
      1470468
    ]
  },
  {
    "key":643,
    "values":[
      329511,
      331997,
      334182,
      335042,
      336741,
      340372
    ]
  }
]
```

**Comment by joaopastor** [comment]
question: is the [execution order idea](https://github.com/schibsted/jslt/blob/master/ideas/group-by.md#execution-order) implemented? by looking at the experimental mode it does not seem so.

**Comment by joaopastor** [comment]
I'd like the elements in each group to be ordered by some criteria (it could be another field, as provided in the input example)
the output for previous example input would be:
```json
[
  {
    "key":642,
    "values":[
      334029,
      335042,
      335156
    ]
  },
  {
    "key":643,
    "values":[
      331997,
      329511,
      334182
    ]
  }
]
```

**Comment by larsga** [comment]
Well, I guess you could do something like this:

```
import "http://jslt.schibsted.com/2018/experimental" as exp
let groups = exp:group-by(., .key, {"value" : .value, "group-order" : .group-order})

[for ($groups)
  {"key" : .key,
   [for (.values) .value order by .group-order]
]
```

That I think would do what you want, even if it's a little convoluted. So this is basically issue #172, although with `order by` instead of `sort()`, but if we add one then adding the other is easy.

**Comment by larsga** [comment]
The `order by` is not implemented yet, no. Sorry. You could work around it by writing a `sort` function and plugging it in.

**Comment by joaopastor** [comment]
I've created an order-by macro in Experimental module.

```java
public static class OrderBy extends AbstractCallable implements Macro {

    public OrderBy() {
      super("order-by", 2, 2);
    }

    public JsonNode call(final Scope scope, JsonNode input,
                         final ExpressionNode[] parameters) {
      // this has to be a macro, because the second argument needs to be
      // evaluated in a special context

      // first find the array that we iterate over
      JsonNode array = parameters[0].apply(scope, input);
      if (array.isNull())
        return NullNode.instance;
      else if (array.isObject())
        array = NodeUtils.convertObjectToArray(array);
      else if (!array.isArray())
        throw new JsltException("Can't order-by on " + array);

      // now start ordering
      List<JsonNode> orderedList = new ArrayList<JsonNode>();
      for (int i = 0; i < array.size(); i++) {
        orderedList.add(array.get(i));
      }

      Collections.sort(orderedList, new Comparator<JsonNode>() {
        public int compare(JsonNode o1, JsonNode o2) {
          JsonNode leftVal = parameters[1].apply(scope, o1);
          JsonNode rightVal = parameters[1].apply(scope, o2);
          Comparable c1 = getComparableFromJsonNode(leftVal);
          Comparable c2 = getComparableFromJsonNode(rightVal);
          if(c1 == null && c2 == null)  return 0;
          if(c1 == null)  return 1;
          if(c2 == null)  return -1;
          return c1.compareTo(c2);
        }
      });

      // ordering is done, build JSON output
      ArrayNode result = NodeUtils.mapper.createArrayNode();
      for (JsonNode val : orderedList) {
        result.add(val);
      }

      return result;
    }
  }

  private static Comparable getComparableFromJsonNode(JsonNode node) {
    Comparable result = null;
    if(!node.isNull()) {
      if(node.isInt() || node.isShort())  result = node.asInt();
      else if(node.isLong())   result = node.asLong();
      else if(node.isDouble() || node.isFloat() || node.isNumber())  result = node.asDouble();
      else if(node.isTextual())   result = node.asText();
      else result = node.asText();
    }
    return result;
  }
```

it uses 2 arguments, the array to ordering criteria accessor. Only works for primitive types.

`exp:order-by(.values, .group_order)`

**Comment by joaopastor** [comment]
Let me know if you want me to create a proper issue and to create a pull-request.

I'd like your assistance in the build process. I cannot get a build with my enhanced code into my local maven repo. Do you have any idea why?

**Comment by joaopastor** [comment]
could this be done as a more pluggable extension?

**Comment by larsga** [comment]
We have an issue for sorting already (issue #172), and implementation is not hard. The difficult part here is defining the sort order clearly, across all the various types. It can be done, but it takes a bit of thinking to get it right.
